lib/core/test/unit/Cardano/Wallet/Primitive/ScriptsSpec.hs

```diff
@@ -0,0 +1,383 @@
+    scriptKeyHashesInMap === Just (Set.fromList (L.nub sciptKeyHashes))
+
+prop_scriptDiscoveredTwice
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptDiscoveredTwice (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let seqState = initializeState accXPub'
+    let (_, seqState') = isShared script seqState
+    let (_, seqState'') = isShared script seqState'
+    getKnownScripts seqState' === getKnownScripts seqState''
+
+prop_scriptsDiscovered
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptsDiscovered (AccountXPubWithScripts accXPub' scripts') = do
+    let seqState0 = initializeState accXPub'
+    let seqState = foldr (\script s -> snd $ isShared script s) seqState0 scripts'
+    let scriptHashes = Set.fromList $ Map.keys $ getKnownScripts seqState
+    scriptHashes === Set.fromList (map toScriptHash scripts')
+
+prop_markingDiscoveredVerKeys
+    :: AccountXPubWithScripts
+    -> Property
+prop_markingDiscoveredVerKeys (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let sciptKeyHashes = retrieveAllVerKeyHashes script
+    let seqState = initializeState accXPub'
+    let (_, SeqState _ _ _ _ _ verKeyPool) =
+            isShared script seqState
+    let ourKeys = verPoolIndexedKeys verKeyPool
+    let discoveredKeyMap =
+            Map.filterWithKey (\k _ -> k `elem` sciptKeyHashes) ourKeys
+    let addressStatesToCheck =
+            map (\(_, (_, isUsed)) -> isUsed) $ Map.toList discoveredKeyMap
+    L.all (== Used) addressStatesToCheck === True
+
+prop_poolExtension
+    :: AccountXPubWithScriptExtension
+    -> Property
+prop_poolExtension (AccountXPubWithScriptExtension accXPub' scripts') = do
+    scriptHashes === Set.fromList (map toScriptHash scripts')
+  where
+    seqState0 = initializeState accXPub'
+    [script1,script2] = scripts'
+    seqState1 = snd $ isShared script1 seqState0
+    seqState2 = snd $ isShared script2 seqState1
+    scriptHashes = Set.fromList $ Map.keys $ getKnownScripts seqState2
+
+prop_scriptsOutsideNotDiscovered
+    :: AccountXPubWithScriptBeyond
+    -> Property
+prop_scriptsOutsideNotDiscovered (AccountXPubWithScriptBeyond accXPub' script) = do
+    let seqState0 = initializeState accXPub'
+    let seqState = snd $ isShared script seqState0
+    let scriptHashes = Map.keys $ getKnownScripts seqState
+    null scriptHashes === True
+
+prop_unusedVerKeysConstant
+    :: AccountXPubWithScripts
+    -> Property
+prop_unusedVerKeysConstant (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let seqState = initializeState accXPub'
+    let (_, seqState') = isShared script seqState
+    let lastUnusedKeys =
+            L.takeWhile (\(_,isUsed) -> isUsed == Unused) .
+            L.sortOn (Down . fst) .
+            Map.elems .
+            getVerKeyMap
+    L.length (lastUnusedKeys seqState) === L.length (lastUnusedKeys seqState')
```

Comment:
It'd be nice to control that for "any" number of scripts traversals, and not a single one. Yet, I'd rather see more tests on the primitives from the Sequential module for that using only `VerificationKeyPools` :thinking:

Reply:
yes, I agree. we can add more tests in Sequential module